silx/gui/plot/actions/control.py

```diff
@@ -632,3 +632,73 @@ def _modeChanged(self, source):
 
     def _actionTriggered(self, checked=False):
         self.plot._eventHandler.validate()
+
+
+class OpenGlAction(PlotAction):
+    """QAction controlling rendering of a :class:`.PlotWidget`.
+
+    For now it can enable or not the OpenGL backend.
+
+    :param plot: :class:`.PlotWidget` instance on which to operate
+    :param parent: See :class:`QAction`
+    """
+
+    def __init__(self, plot, parent=None):
+        # Uses two images for checked/unchecked states
+        self._states = {
+            "opengl": (icons.getQIcon('backend-opengl'),
+                       "OpenGL rendering (fast)\nClick to disable OpenGL"),
+            "matplotlib": (icons.getQIcon('backend-opengl-disabled'),
+                          "Matplotlib rendering (safe)\nClick to enable OpenGL"),
+            "unknown": (icons.getQIcon('backend-opengl-disabled'),
+                        "Custom rendering")
+        }
+
+        name = self._getBackendName(plot)
+        self.__state = name
+        icon, tooltip = self._states[name]
+        super(OpenGlAction, self).__init__(
+            plot,
+            icon=icon,
+            text='Toggle keep aspect ratio',
+            tooltip=tooltip,
+            triggered=self._actionTriggered,
+            checkable=False,
+            parent=parent)
+
+    def _backendUpdated(self):
+        name = self._getBackendName(self.plot)
+        self.__state = name
+        icon, tooltip = self._states[name]
+        self.setIcon(icon)
+        self.setToolTip(tooltip)
+
+    def _getBackendName(self, plot):
+        backend = plot.getBackend()
+        name = type(backend).__name__.lower()
+        if "opengl" in name:
+            return "opengl"
+        elif "matplotlib" in name:
+            return "matplotlib"
+        else:
+            return "unknown"
+
+    def _actionTriggered(self, checked=False):
+        plot = self.plot
+        name = self._getBackendName(self.plot)
+        if self.__state != name:
+            # THere is no event to know the backend was updated
+            # So here we check if there is a mismatch between the displayed state
+            # and the real state of the widget
+            self._backendUpdated()
+            return
+        if name is not "opengl":
+            from silx.gui.utils import glutils
+            result = glutils.isOpenGLAvailable()
```

Comment:
I would make the test at action creation, but I guess you want to keep it fast.
Once it failed once, you can disable the action, it will never become available.

Reply:
Yes i don't want crazy tests at creation.

I don't know about disabling anyway. I prefer not to do so.

If you click on it again you have an obvious error message, you don't need to make it safer.
If you are able to install `pyopengl` in your env, the problem could be fixed without restarting the application.

Reply:
It would work for install `pyopengl`, but for any other error, it would not since `isOpenGLAvailable` caches the runtime check result.

Reply:
That's right, we could add a parameter to skip the cache here.

Reply:
Well, there is no reason the result of the runtime check would change...
`pyopengl` availability, `$DISPLAY` and Qt reporting an issue are checked each time, only the runtime check which is run after those first checks is cached.

Reply:
Driver update, or lib update

Reply:
without restarting the app, things will already be loaded, no?